src/lib.rs

```diff
@@ -457,34 +520,224 @@ impl Cascade {
+    ) -> Self {
+        CascadeBuilder {
+            filters: vec![Bloom::new_crlite_bloom(
+                include_capacity,
+                exclude_capacity,
+                true,
+            )],
+            salt,
+            to_include: vec![],
+            to_exclude: vec![],
+            hash_algorithm,
+            status: BuildStatus(include_capacity, exclude_capacity),
+        }
+    }
+
+    pub fn include(&mut self, item: Vec<u8>) {
+        match self.status {
+            BuildStatus(ref mut cap, _) if *cap > 0 => *cap -= 1,
+            _ => panic!("capacity violation"),
+        }
+        let mut generator = CascadeIndexGenerator::new(self.hash_algorithm, item);
+        self.filters[0].insert(&mut generator, &self.salt);
+        self.to_include.push(generator);
+    }
+
+    pub fn exclude(&mut self, item: Vec<u8>) {
+        match self.status {
+            BuildStatus(0, ref mut cap) if *cap > 0 => *cap -= 1,
+            _ => panic!("capacity violation"),
+        }
+        let mut generator = CascadeIndexGenerator::new(self.hash_algorithm, item);
+        if self.filters[0].has(&mut generator, &self.salt) {
+            self.to_exclude.push(generator);
+        }
+    }
+
+    /// `exclude_threaded` is like `exclude` but it stores false positives in a caller-owned
+    /// `ExcludeSet`. This allows the caller to exclude items in parallel.
+    pub fn exclude_threaded(&self, exclude_set: &mut ExcludeSet, item: Vec<u8>) {
+        exclude_set.size += 1;
+        let mut generator = CascadeIndexGenerator::new(self.hash_algorithm, item);
+        if self.filters[0].has(&mut generator, &self.salt) {
+            exclude_set.set.push(generator);
+        }
+    }
+
+    /// `collect_exclude_set` merges an `ExcludeSet` into the internal storage of the CascadeBuilder.
+    pub fn collect_exclude_set(&mut self, exclude_set: &mut ExcludeSet) {
+        match self.status {
+            BuildStatus(0, ref mut cap) if *cap >= exclude_set.size => *cap -= exclude_set.size,
+            _ => panic!("capacity violation"),
+        }
+        self.to_exclude.append(&mut exclude_set.set);
+    }
+
+    fn push_layer(&mut self) -> Result<(), CascadeError> {
+        // At even layers we encode elements of to_include. At odd layers we encode elements of
+        // to_exclude. In both cases, we track false positives by filtering the complement of the
+        // encoded set through the newly produced bloom filter.
+        let at_even_layer = self.filters.len() % 2 == 0;
+        let (to_encode, to_filter) = match at_even_layer {
+            true => (&mut self.to_include, &mut self.to_exclude),
+            false => (&mut self.to_exclude, &mut self.to_include),
+        };
+
+        let mut bloom = Bloom::new_crlite_bloom(to_encode.len(), to_filter.len(), false);
+
+        // temporarily take self.salt since we need an immutable reference for bloom.insert
+        let salt = std::mem::take(&mut self.salt);
```

Comment:
Since `self.salt` is no longer an `Option`, we should be able to use `as_slice()`, right?

Reply:
Let me know if you see a trick, but I don't see a way around the `take` here. We have to split a borrow of `self` into a mutable borrow of `self.to_filter` and an immutable borrow of `self.salt`. The borrow checker is smart enough to do this in a for loop, but it doesn't like when we capture `self.salt` in a closure and then try to pass mutable `self.to_filter` elements to that closure in `retain_mut`.

Reply:
It looks like creating the slice outside the loop/closure works: `let salt_slice = self.salt.as_slice();`

Reply:
Oh, thanks!